contracts/VetoMultisigVoting.sol

```diff
@@ -0,0 +1,243 @@
+import "./interfaces/IVetoGuard.sol";
+import "./interfaces/IVetoVoting.sol";
+import "./interfaces/IGnosisSafe.sol";
+import "./TransactionHasher.sol";
+import "@gnosis.pm/safe-contracts/contracts/common/Enum.sol";
+import "@gnosis.pm/zodiac/contracts/factory/FactoryFriendly.sol";
+
+/// @notice A contract for a parent Multisig DAO to cast veto and freeze votes on a child DAO
+contract VetoMultisigVoting is IVetoVoting, TransactionHasher, FactoryFriendly {
+    uint256 public vetoVotesThreshold; // Number of votes required to veto a transaction
+    uint256 public freezeVotesThreshold; // Number of freeze votes required to activate a freeze
+    uint256 public freezeProposalCreatedBlock; // Block number the freeze proposal was created at
+    uint256 public freezeProposalVoteCount; // Number of accrued freeze votes
+    uint256 public freezeProposalBlockDuration; // Number of blocks a freeze proposal has to succeed
+    uint256 public freezeBlockDuration; // Number of blocks a freeze lasts, from time of freeze proposal creation
+    IGnosisSafe public gnosisSafe;
+    IVetoGuard public vetoGuard;
+    mapping(bytes32 => uint256) public transactionVetoVotes;
+    mapping(address => mapping(bytes32 => bool)) public userHasVoted;
+    mapping(address => mapping(uint256 => bool)) public userHasFreezeVoted;
+
+    /// @notice Initialize function, will be triggered when a new proxy is deployed
+    /// @param initializeParams Parameters of initialization encoded
+    function setUp(bytes memory initializeParams) public override initializer {
+        __Ownable_init();
+        (
+            address _owner,
+            uint256 _vetoVotesThreshold,
+            uint256 _freezeVotesThreshold,
+            uint256 _freezeProposalBlockDuration,
+            uint256 _freezeBlockDuration,
+            address _parentGnosisSafe,
+            address _vetoGuard
+        ) = abi.decode(
+                initializeParams,
+                (address, uint256, uint256, uint256, uint256, address, address)
+            );
+
+        _transferOwnership(_owner);
+        vetoVotesThreshold = _vetoVotesThreshold;
+        freezeVotesThreshold = _freezeVotesThreshold;
+        freezeProposalBlockDuration = _freezeProposalBlockDuration;
+        freezeBlockDuration = _freezeBlockDuration;
+        gnosisSafe = IGnosisSafe(_parentGnosisSafe);
+        vetoGuard = IVetoGuard(_vetoGuard);
+    }
+
+    /// @notice Allows the msg.sender to cast veto and freeze votes on the specified transaction
+    /// @param _transactionHash The hash of the transaction data
+    /// @param _freeze Bool indicating whether the voter thinks the DAO should also be frozen
+    function castVetoVote(bytes32 _transactionHash, bool _freeze) external {
+        // Check that user has not yet voted
+        require(
+            !userHasVoted[msg.sender][_transactionHash],
+            "User has already voted"
+        );
+
+        require(gnosisSafe.isOwner(msg.sender), "User is not an owner ");
```

Comment:
Could save a non owner a little bit of gas if this is first within the requirement block.

Reply:
But then it just hits it before the other require statement, and I could make the same argument that a user is more likely to try to vote twice than a non-owner is to try and cast a veto vote.